Cpp/nBodySimulator/config.txt
//INSTRUCTIONS:
//Separate key = value pairs with spaces!
//Don't use spaces for anything else (use _ instead)
//
//Positions/Velocities are RELATIVE
//i.e.
//positions of planets are relative to the to the parent solarsystem
//positions of solarsystems are relative to the parent galaxy
//etc.
//This makes it very easy to create a lot of systems


[settings]
//how many times should the screen update each second
FRAMES_PER_SECOND = 60
//how many seconds should be simulated each time the screen is redrawn
PHYSICS_STEP_SIZE = 40320 // 1 lunar month per 60 frames
//split each physics step into this many sub-steps
OVERCALCULATE_PHYSICS = 10
//meters per pixel, between bodies
DISTANCE_SCALE = 1.0e10
//meters per pixel, for body radius
RADIUS_SCALE = 1.0e7
//use a log2 scale for object radius, instead of RADIUS_SCALE?
USE_LOG_RADIUS = 0
//if using log2 scale, what radius in m is ~1 pixel?
LOG_POINT = 1.0e6


[zodiac]
Alpha_Centauri_A = Alpha_Centauri_B
Alpha_Centauri_A = Proxima_Centauri
Alpha_Centauri_B = Proxima_Centauri

[zodiac]
Sun = Sun_B
Sun = Sun_C
Earth = Sun
Earth_B = Sun_B
Earth_C = Sun_C

[blackhole]
name = EaterOfWorlds
position_x = 3.0e12
position_y = 3.0e12
velocity_x = 0.0
velocity_y = 0.0
radius = 1.0e8
mass = 1.0e32
color = #440044
parent      =  Milky_Way

[cluster]
name = Cluster_A

[cluster]
name = Cluster_B
parent = Cluster_A

[galaxy]
name = Milky_Way
position_x  =  0
position_y  =  0
velocity_x  =  0
velocity_y  =  0
parent = Cluster_A

[galaxy]
name = Milky_Way_B
position_x  =  0
position_y  =  0
velocity_x  =  0
velocity_y  =  0
parent = Cluster_A

[solarsystem]
name = Sol
position_x  =  0
position_y  =  0
velocity_x  =  0
velocity_y  =  0
parent = Milky_Way

[solarsystem]
name = Sol_B
position_x  =  1.0e12
position_y  =  0
velocity_x  =  0
velocity_y  =  1.0e4
parent = Milky_Way_B

[solarsystem]
name = Sol_C
position_x  =  -1.0e12
position_y  =  -2.0e12
velocity_x  =  0
velocity_y  =  3.0e4
parent = Milky_Way_B


[star]
name = Sun
position_x = 0
position_y = 0
velocity_x = 0
velocity_y = 0
radius = 696e6
mass = 1.909e30
color = #FFFF00
parent      =  Sol

[planet]
name        =  Mercury
position_x  =  46.0e9
position_y  =  0
velocity_x  =  0
velocity_y  =  58.98e3
radius      =  2.44e6
mass        =  0.330e24
color       =  #808080   //grey
parent      =  Sol

[planet]
name        = Venus
position_x  = 107.5e9
position_y  = 0
velocity_x  = 0
velocity_y  = 35.26e3
radius      = 6.052e6
mass        = 4.87e24
color       = #FFFF40    //yellow-white
parent      =  Sol

[planet]
name = Earth
position_x = 147.09e9
position_y = 0
velocity_x = 0
velocity_y = 30.29e3
radius = 6.378e6
mass = 5.972e24
color = #0040FF //blue-green
parent = Sol

[planet]
name = Moon
position_x = 147.473e9
position_y = 0
velocity_x = 0
velocity_y = 31.366e3
radius = 1.738e6
mass = 0.073e24
color = #808080 //grey
parent      =  Sol

[planet]
name = Mars
position_x = 206.62e9
position_y = 0
velocity_x = 0
velocity_y = 26.50e3
radius = 1.962e6
mass = 0.64171e24
color = #FF4020 //red-brown
parent      =  Sol

[planet]
name = Jupiter
position_x = 740.5e9
position_y = 0
velocity_x = 0
velocity_y = 13.72e3
radius = 71.492e6
mass = 1898e24
color = #A04040 //mix, browns, reds, blues, whites
parent      =  Sol

[planet]
name = Saturn
position_x = 1352.6e9
position_y = 0
velocity_x = 0
velocity_y = 10.18e3
radius = 60.268e6
mass = 568e24
color = #806040 //mix
parent      =  Sol

[planet]
name = Uranus
position_x = 2741.3e9
position_y = 0
velocity_x = 0
velocity_y = 7.11e3
radius = 25.559e6
mass = 86.8e24
color = #4080FF //light blue-green
parent      =  Sol

[planet]
name = Neptune
position_x = 4444.5e9
position_y = 0
velocity_x = 0
velocity_y = 5.5e3
radius = 24.764e6
mass = 102e24
color = #2060FF //blue-green
parent      =  Sol

[planet]
name = Pluto
position_x = 4436.8e9
position_y = 0
velocity_x = 0
velocity_y = 6.1e3
radius = 1.185e6
mass = 0.0146e24
color = #FF6600 //dark orange
parent      =  Sol



[solarsystem]
name = Alpha_Centauri
position_x  =  3.0e4
position_y  =  3.0e16
velocity_x  =  0
velocity_y  =  0
parent = Milky_Way

[star]
name = Alpha_Centauri_A
position_x = 0
position_y = 0
velocity_x = 0
velocity_y = 0
radius = 854e6
mass = 2.1e30
color = #FFFF22
parent      =  Alpha_Centauri

[star]
name = Alpha_Centauri_B
position_x = 1.6e12
position_y = 0
velocity_x = 0
velocity_y = 7.0e4
radius = 696e6
mass = 1.909e30
color = #FFFF00
parent      =  Alpha_Centauri

[star]
name = Proxima_Centauri
position_x = 0
position_y = 2.0e15
velocity_x = 0
velocity_y = 0
radius = 1.0e8
mass = 2.1e29
color = #AA0000
parent      =  Alpha_Centauri





[star]
name = Sun_B
position_x = 0
position_y = 0
velocity_x = 0
velocity_y = 0
radius = 696e6
mass = 1.909e30
color = #FFFF00
parent      =  Sol_B

[planet]
name        =  Mercury_B
position_x  =  46.0e9
position_y  =  0
velocity_x  =  0
velocity_y  =  58.98e3
radius      =  2.44e6
mass        =  0.330e24
color       =  #808080   //grey
parent      =  Sol_B

[planet]
name        = Venus_B
position_x  = 107.5e9
position_y  = 0
velocity_x  = 0
velocity_y  = 35.26e3
radius      = 6.052e6
mass        = 4.87e24
color       = #FFFF40    //yellow-white
parent      =  Sol_B

[planet]
name = Earth_B
position_x = 147.09e9
position_y = 0
velocity_x = 0
velocity_y = 30.29e3
radius = 6.378e6
mass = 5.972e24
color = #0040FF //blue-green
parent = Sol_B

[planet]
name = Moon_B
position_x = 147.473e9
position_y = 0
velocity_x = 0
velocity_y = 31.366e3
radius = 1.738e6
mass = 0.073e24
color = #808080 //grey
parent      =  Sol_B

[planet]
name = Mars_B
position_x = 206.62e9
position_y = 0
velocity_x = 0
velocity_y = 26.50e3
radius = 1.962e6
mass = 0.64171e24
color = #FF4020 //red-brown
parent      =  Sol_B

[planet]
name = Jupiter_B
position_x = 740.5e9
position_y = 0
velocity_x = 0
velocity_y = 13.72e3
radius = 71.492e6
mass = 1898e24
color = #A04040 //mix, browns, reds, blues, whites
parent      =  Sol_B

[planet]
name = Saturn_B
position_x = 1352.6e9
position_y = 0
velocity_x = 0
velocity_y = 10.18e3
radius = 60.268e6
mass = 568e24
color = #806040 //mix
parent      =  Sol_B

[planet]
name = Uranus_B
position_x = 2741.3e9
position_y = 0
velocity_x = 0
velocity_y = 7.11e3
radius = 25.559e6
mass = 86.8e24
color = #4080FF //light blue-green
parent      =  Sol_B

[planet]
name = Neptune_B
position_x = 4444.5e9
position_y = 0
velocity_x = 0
velocity_y = 5.5e3
radius = 24.764e6
mass = 102e24
color = #2060FF //blue-green
parent      =  Sol_B

[planet]
name = Pluto_B
position_x = 4436.8e9
position_y = 0
velocity_x = 0
velocity_y = 6.1e3
radius = 1.185e6
mass = 0.0146e24
color = #FF6600 //dark orange
parent      =  Sol_B





[star]
name = Sun_C
position_x = 0
position_y = 0
velocity_x = 0
velocity_y = 0
radius = 696e6
mass = 1.909e30
color = #FFFF00
parent      =  Sol_C

[planet]
name        =  Mercury_C
position_x  =  -46.0e9
position_y  =  0
velocity_x  =  0
velocity_y  =  58.98e3
radius      =  2.44e6
mass        =  0.330e24
color       =  #808080   //grey
parent      =  Sol_C

[planet]
name        = Venus_C
position_x  = -107.5e9
position_y  = 0
velocity_x  = 0
velocity_y  = 35.26e3
radius      = 6.052e6
mass        = 4.87e24
color       = #FFFF40    //yellow-white
parent      =  Sol_C

[planet]
name = Earth_C
position_x = -147.09e9
position_y = 0
velocity_x = 0
velocity_y = 30.29e3
radius = 6.378e6
mass = 5.972e24
color = #0040FF //blue-green
parent = Sol_C

[planet]
name = Moon_C
position_x = -147.473e9
position_y = 0
velocity_x = 0
velocity_y = 31.366e3
radius = 1.738e6
mass = 0.073e24
color = #808080 //grey
parent      =  Sol_C

[planet]
name = Mars_C
position_x = -206.62e9
position_y = 0
velocity_x = 0
velocity_y = 26.50e3
radius = 1.962e6
mass = 0.64171e24
color = #FF4020 //red-brown
parent      =  Sol_C

[planet]
name = Jupiter_C
position_x = -740.5e9
position_y = 0
velocity_x = 0
velocity_y = 13.72e3
radius = 71.492e6
mass = 1898e24
color = #A04040 //mix, browns, reds, blues, whites
parent      =  Sol_C

[planet]
name = Saturn_C
position_x = -1352.6e9
position_y = 0
velocity_x = 0
velocity_y = 10.18e3
radius = 60.268e6
mass = 568e24
color = #806040 //mix
parent      =  Sol_C

[planet]
name = Uranus_C
position_x = -2741.3e9
position_y = 0
velocity_x = 0
velocity_y = 7.11e3
radius = 25.559e6
mass = 86.8e24
color = #4080FF //light blue-green
parent      =  Sol_C

[planet]
name = Neptune_C
position_x = -4444.5e9
position_y = 0
velocity_x = 0
velocity_y = 5.5e3
radius = 24.764e6
mass = 102e24
color = #2060FF //blue-green
parent      =  Sol_C

[planet]
name = Pluto_C
position_x = -4436.8e9
position_y = 0
velocity_x = 0
velocity_y = 6.1e3
radius = 1.185e6
mass = 0.0146e24
color = #FF6600 //dark orange
parent      =  Sol_C



Cpp/nBodySimulator/config.cpp
#include "config.h"
#include "universecomposite.h"
#include "universecomponentfactory.h"
#include <QDebug>
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <utility>
#include <sstream>

using std::string;
using std::list;
using std::ifstream;
using std::stringstream;
using std::exception;
using std::invalid_argument;
using std::out_of_range;

//Singleton variables
bool Config::s_initialised = false;
Config* Config::s_instance = nullptr;


//a type to make it easier to keep track of what the parser is doing
enum BlockType { none, settings, component, zodiac };

Config::Config() // Initialize to default values
    : m_framesPerSecond(60)
    , m_physicsStepSize(40320)
    , m_overcalculatePhysics(1)
    , m_distanceScale(1.0e10)
    , m_radiusScale(1.0e7)
    , m_useLogRadius(true)
    , m_logPointRadius(1.0e6)
    , m_stepSizeChange(0)
    , m_distanceScaleChange(0)
    , m_radiusScaleChange(0)
    , m_logPointRadiusChange(0)
{
}

Config::~Config()
{
    //assigning responsibility for deleting the universe to the caller (Dialog)
    //using shared_pointer would be a much smarter solution
}

Config* Config::getInstance()
{
    if(!s_initialised)
    {
        s_instance = new Config();
        s_initialised = true;
    }
    return s_instance;
}

/***** Text content processing/checking *****/
string Config::stripComment(const string& line) const
{
    if(line.length() == 0)
    {
        return line;
    }
    int index = line.find_first_of('/');
    return line.substr(0, index);
}

bool Config::isBlockHeading(const string& line) const
{
    return (line.length() > 0)
            && (line[0] == '[')
            && (line[line.length()-1] == ']');
}

bool Config::isKeyValuePair(const string& key, const string& op, const string& value) const
{
    return (key.length() > 0) && (op == "=") && (value.length() > 0);
}

//acquire file, read strings, close file
void Config::readLines(const string& path, list<string>& output) const
{
    ifstream configfile(path.c_str());
    if(!configfile.is_open()) {
        qDebug() << "Unable to open file: " << path.c_str();
    }
    string line;
    while(getline(configfile, line))
    {
        output.push_back(line);
    }
    configfile.close();
}

void Config::read(const string& path)
{
    m_settings.clear();
    m_universeBlocks.clear();
    m_zodiacBlocks.clear();

    list<string> lines;
    std::unordered_map<string, string> block;
    list<std::pair<string, string> > zodiacBlock;

    BlockType currentBlockType = none;

    readLines(path, lines);
    int lineNumber = 0; // for debugging purpose

    for(string line : lines)
    {
        ++lineNumber;

        //try to read three strings (breaks on spaces!)
        stringstream ss(stripComment(line));
        string key, op, value;
        ss >> key >> op >> value;

        //skip blank lines silently
        if(key.length() == 0) continue;

        //starting a new heading
        if(isBlockHeading(key)) {
            //save the finished block
            switch(currentBlockType) {
            case component:
                m_universeBlocks.emplace_back(block);
                block.clear();
                break;
            case zodiac:
                m_zodiacBlocks.emplace_back(zodiacBlock);
                zodiacBlock.clear();
                break;
            case settings:
            case none:
                break;
            }

            //start the new block
            string label = key.substr(1, key.length()-2);
            if(label == "zodiac") {
                currentBlockType = zodiac;
            } else if(label == "settings") {
                currentBlockType = settings;
            } else {
                currentBlockType = component;
                block.clear();
                block.insert(make_pair("category", label)); // category : planet/cluster/galaxy/blackhole....
            }
        }
        //read a k-v pair within a heading
        else if(isKeyValuePair(key, op, value)) {
            switch(currentBlockType) {
            case settings:
                m_settings.insert(make_pair(key, value));
                break;
            case component:
                block.insert(make_pair(key, value));
                break;
            case zodiac:
                zodiacBlock.emplace_back(make_pair(key, value));
                break;
            case none:
            default:
                qDebug() << "ERROR on line" << lineNumber << ", file should have started with a [heading]";
            }
        }
        else {
            qDebug() << "ERROR on line" << lineNumber
                     << "Expected: 'key = value' or '[block]'"
                     << "actually:"
                     << line.c_str();
        }
    }

    //end of file, save the finished block
    switch(currentBlockType) {
    case component:
        m_universeBlocks.emplace_back(block);
        block.clear();
        break;
    case zodiac:
        m_zodiacBlocks.emplace_back(zodiacBlock);
        zodiacBlock.clear();
        break;
    case settings:
    case none:
        break;
    }

    //update the settings
    parseSettings();
}

void Config::parseSettings()
{
    if (m_settings.count("FRAMES_PER_SECOND") > 0) {
        stringstream ss(m_settings.at("FRAMES_PER_SECOND"));
        ss >> m_framesPerSecond;
        qDebug() << "FRAMES_PER_SECOND" << m_framesPerSecond;
    }

    if (m_settings.count("PHYSICS_STEP_SIZE") > 0) {
        stringstream ss(m_settings.at("PHYSICS_STEP_SIZE"));
        ss >> m_physicsStepSize;
        qDebug() << "PHYSICS_STEP_SIZE" << m_physicsStepSize;
    }

    if (m_settings.count("OVERCALCULATE_PHYSICS") > 0) {
        stringstream ss(m_settings.at("OVERCALCULATE_PHYSICS"));
        ss >> m_overcalculatePhysics;
        if(m_overcalculatePhysics < 1)
            m_overcalculatePhysics = 1;
        qDebug() << "OVERCALCULATE_PHYSICS" << m_overcalculatePhysics;
    }

    if (m_settings.count("DISTANCE_SCALE") > 0) {
        stringstream ss(m_settings.at("DISTANCE_SCALE"));
        ss >> m_distanceScale;
        qDebug() << "DISTANCE_SCALE" << m_distanceScale;
    }

    if (m_settings.count("RADIUS_SCALE") > 0) {
        stringstream ss(m_settings.at("RADIUS_SCALE"));
        ss >> m_radiusScale;
        qDebug() << "RADIUS_SCALE" << m_radiusScale;
    }

    if (m_settings.count("USE_LOG_RADIUS") > 0) {
        stringstream ss(m_settings.at("USE_LOG_RADIUS"));
        ss >> m_useLogRadius;
        qDebug() << "USE_LOG_RADIUS" << m_useLogRadius;
    }

    if (m_settings.count("LOG_POINT") > 0) {
        stringstream ss(m_settings.at("LOG_POINT"));
        ss >> m_logPointRadius;
        qDebug() << "LOG_POINT" << m_logPointRadius;
    }
}

void Config::joinComponents(UniverseComponent* parent, UniverseComponent* child)
{
    switch(parent->getType()) {
    //allow anything to join to the universe
    case universe:
        dynamic_cast<UniverseComposite*>(parent)->add(child);
        return;

    //allow clusters and galaxies to join to a cluster
    case cluster:
        switch(child->getType()) {
        case cluster:
        case galaxy:
            dynamic_cast<UniverseComposite*>(parent)->add(child);
            return;
        default:
            break;
        }
        break;

    //allow solarsystems and blackholes to join to a galaxy
    case galaxy:
        switch(child->getType()) {
        case solarsystem:
        case blackhole:
            dynamic_cast<UniverseComposite*>(parent)->add(child);
            return;
        default:
            break;
        }
        break;

    //only allow bodies to join to a solarsystem
    case solarsystem:
        switch(child->getType()) {
        case star:
        case blackhole:
        case planet:
            dynamic_cast<UniverseComposite*>(parent)->add(child);
            return;
        default:
            break;
        }
        break;
    default:
        break;
    }

    //anything else is invalid
    qDebug() << "ERROR: cannot add ("
             << child->getName().c_str()
             << ") to ("
             << parent->getName().c_str()
             << "). They are the wrong types. Skipping object.";
    return;

}


UniverseComponent* Config::parseUniverseBlocks()
{
    UniverseComponentFactory factory;
    UniverseComponent* universe = new UniverseComposite(
                UniverseComponentType::universe, "The Universe");

    m_universeComponents.clear();

    //create the components
    for(auto block : m_universeBlocks) {
        try {
            UniverseComponent* component = factory.createUniverseComponent(block);
            if(m_universeComponents.count(component->getName()) > 0)
                qDebug() << "ERROR: name already in use (skipping):" << component->getName().c_str();
            else
                m_universeComponents.insert(make_pair(component->getName(), component));
        }
        catch(invalid_argument e) {
            qDebug() << "ERROR: skipped an object:" << e.what();
        }
        catch(out_of_range e) {
            qDebug() << "ERROR: skipped an object:" << e.what();
        }
        catch(exception e) {
            qDebug() << "ERROR: skipped an object:" << e.what();
        }
    }

    //link the components to their parents (or to the Universe)
    for (auto pair : m_universeComponents) {
        UniverseComponent* child = pair.second;
        UniverseComponent* parent = universe;
        if (child->getParentName() != "") {
            parent = m_universeComponents.at(child->getParentName());
        }
        joinComponents(parent, child);
    }

    return universe;
}

void Config::addToZodiac(const string& labelA, const string& labelB, Zodiac& zodiac)
{
    //find the first object, and make sure it's a suitable type
    if(m_universeComponents.count(labelA) < 1) {
        qDebug() << "ERROR: couldn't find" << labelA.c_str()
                 << "for zodiac, skipping line";
        return;
    }
    UniverseComponent* starA = m_universeComponents.at(labelA);

    switch(starA->getType()) {
    case star:
    case blackhole:
    case planet:
        break;
    default:
        qDebug() << "ERROR:" << labelA.c_str()
                 << "is not a body (for zodiac), skipping line";
        return ;
    }

    //find the second object, and make sure it's a suitable type
    if(m_universeComponents.count(labelB) < 1) {
        qDebug() << "ERROR: couldn't find" << labelB.c_str()
                 << "for zodiac, skipping line";
        return;
    }

    UniverseComponent* starB = m_universeComponents.at(labelB);
    switch(starB->getType()) {
    case star:
    case blackhole:
    case planet:
        break;
    default:
        qDebug() << "ERROR:" << labelB.c_str()
                 << "is not a body (for zodiac), skipping line";
        return ;
    }

    //add a line between the two objects, to the zodiac
    zodiac.add(dynamic_cast<UniverseBody*>(starA),
               dynamic_cast<UniverseBody*>(starB));
}


list<Zodiac>* Config::parseZodiacBlocks()
{
    list<Zodiac>* zodiacs = new list<Zodiac>;
    //for every block of zodiac defined in the file
    for(auto block : m_zodiacBlocks) {
        Zodiac zodiac;
        //for every pair of objects in the block
        for(auto pair : block) {
            //add the line
            addToZodiac(pair.first, pair.second, zodiac);
        }
        //add the finished zodiac
        zodiacs->emplace_back(zodiac);
    }
    return zodiacs;
}
